File: mobile/src/services/business.ts
import { apiClient } from '@/lib/api';
import { Business } from '@/types';

export interface BusinessInput {
  name: string;
  owner_name: string;
  phone: string;
  email: string;
  address: string;
  gstin?: string;
  logo_url?: string;
  currency?: string;
}

export interface BusinessUpdateInput {
  name?: string;
  owner_name?: string;
  phone?: string;
  email?: string;
  address?: string;
  gstin?: string;
  logo_url?: string;
  currency?: string;
}

interface ApiBusinessResponse {
  id: string;
  user_id: string;
  name: string;
  owner_name: string;
  phone: string;
  email: string;
  address: string;
  gstin?: string | null;
  logo_url?: string | null;
  currency: string;
  is_default: boolean;
  created_at: string;
  updated_at: string;
}

function mapApiBusiness(apiBiz: ApiBusinessResponse): Business {
  return {
    id: apiBiz.id,
    userId: apiBiz.user_id,
    name: apiBiz.name,
    ownerName: apiBiz.owner_name,
    phone: apiBiz.phone,
    email: apiBiz.email,
    address: apiBiz.address,
    gstin: apiBiz.gstin ?? undefined,
    logoUrl: apiBiz.logo_url ?? undefined,
    currency: apiBiz.currency,
    isDefault: apiBiz.is_default,
  };
}

export const BusinessService = {
  /**
   * Fetch current authenticated user's business profile.
   * Returns null if 404 (user hasn't completed business setup yet).
   */
  async getMyBusiness(): Promise<Business | null> {
    try {
      const response = await apiClient.get<ApiBusinessResponse>('/api/v1/business/me');
      return mapApiBusiness(response.data);
    } catch (error: unknown) {
      if (typeof error === 'object' && error !== null && 'response' in error) {
        const status = (error as { response?: { status?: number } }).response?.status;
        if (status === 404) {
          return null;
        }
      }
      throw error;
    }
  },

  /**
   * Create initial business profile for the current user.
   */
  async createBusiness(input: BusinessInput): Promise<Business> {
    const response = await apiClient.post<ApiBusinessResponse>('/api/v1/business', input);
    return mapApiBusiness(response.data);
  },

  /**
   * Update current user's business profile.
   */
  async updateBusiness(input: BusinessUpdateInput): Promise<Business> {
    const response = await apiClient.put<ApiBusinessResponse>('/api/v1/business/me', input);
    return mapApiBusiness(response.data);
  },

  /**
   * List all user businesses (supports multi-business).
   */
  async listMyBusinesses(): Promise<Business[]> {
    const response = await apiClient.get<ApiBusinessResponse[]>('/api/v1/business');
    return response.data.map(mapApiBusiness);
  },
};

File: backend/app/api/v1/endpoints/business.py
from typing import List, Optional
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from sqlalchemy import select

from app.dependencies.db import get_db
from app.dependencies.auth import get_current_user
from app.models.user import User
from app.models.business import Business
from app.schemas.business import BusinessCreate, BusinessUpdate, BusinessResponse

router = APIRouter()


@router.get("/me", response_model=BusinessResponse)
def get_my_business(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """
    Retrieve the current active business profile for the authenticated user.
    Returns 404 if the user hasn't created a business profile yet.
    """
    stmt = (
        select(Business)
        .where(Business.user_id == current_user.id)
        .order_by(Business.is_default.desc(), Business.created_at.asc())
    )
    business = db.scalars(stmt).first()
    if not business:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Business profile not found. Please set up your business profile.",
        )
    return business


@router.post("", response_model=BusinessResponse, status_code=status.HTTP_201_CREATED)
def create_business(
    payload: BusinessCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """
    Create a new business profile for the authenticated user.
    Initially supports 1 business per user. If one exists, returns 400.
    """
    stmt = select(Business).where(Business.user_id == current_user.id)
    existing_business = db.scalars(stmt).first()
    if existing_business:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Business profile already exists for this user.",
        )

    business = Business(
        user_id=current_user.id,
        name=payload.name,
        owner_name=payload.owner_name,
        phone=payload.phone,
        email=payload.email,
        address=payload.address,
        gstin=payload.gstin,
        logo_url=payload.logo_url,
        currency=payload.currency,
        invoice_prefix=payload.invoice_prefix or "INV",
        quotation_prefix=payload.quotation_prefix or "QT",
        default_terms=payload.default_terms,
        payment_instructions=payload.payment_instructions,
        is_default=True,
    )
    db.add(business)
    db.commit()
    db.refresh(business)
    return business


@router.put("/me", response_model=BusinessResponse)
def update_my_business(
    payload: BusinessUpdate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """
    Update the active business profile for the authenticated user.
    """
    stmt = (
        select(Business)
        .where(Business.user_id == current_user.id)
        .order_by(Business.is_default.desc(), Business.created_at.asc())
    )
    business = db.scalars(stmt).first()
    if not business:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Business profile not found. Please set up your business profile first.",
        )

    update_data = payload.model_dump(exclude_unset=True)
    for field, value in update_data.items():
        setattr(business, field, value)

    db.commit()
    db.refresh(business)
    return business


@router.get("", response_model=List[BusinessResponse])
def list_my_businesses(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """
    List all businesses associated with the authenticated user (supports future multi-business).
    """
    stmt = (
        select(Business)
        .where(Business.user_id == current_user.id)
        .order_by(Business.is_default.desc(), Business.created_at.asc())
    )
    return list(db.scalars(stmt).all())
